internal/graph/builder.go

```diff
@@ -0,0 +1,84 @@
+package graph
+
+type CheckResolverOrderedBuilder struct {
+	resolvers                              []CheckResolver
+	localCheckerOptions                    []LocalCheckerOption
+	cachedCheckResolverEnabled             bool
+	cachedCheckResolverOptions             []CachedCheckResolverOpt
+	dispatchThrottlingCheckResolverEnabled bool
+	dispatchThrottlingCheckResolverOptions []DispatchThrottlingCheckResolverOpt
+}
+
+type CheckResolverOrderedBuilderOpt func(checkResolver *CheckResolverOrderedBuilder)
+
+// WithLocalCheckerOpts sets the opts to be used to build LocalChecker.
+func WithLocalCheckerOpts(opts ...LocalCheckerOption) CheckResolverOrderedBuilderOpt {
+	return func(r *CheckResolverOrderedBuilder) {
+		r.localCheckerOptions = opts
+	}
+}
+
+// WithCachedCheckResolverOpts sets the opts to be used to build CachedCheckResolver.
+func WithCachedCheckResolverOpts(enabled bool, opts ...CachedCheckResolverOpt) CheckResolverOrderedBuilderOpt {
+	return func(r *CheckResolverOrderedBuilder) {
+		r.cachedCheckResolverEnabled = enabled
+		r.cachedCheckResolverOptions = opts
+	}
+}
+
+// WithDispatchThrottlingCheckResolverOpts sets the opts to be used to build DispatchThrottlingCheckResolver.
+func WithDispatchThrottlingCheckResolverOpts(enabled bool, opts ...DispatchThrottlingCheckResolverOpt) CheckResolverOrderedBuilderOpt {
+	return func(r *CheckResolverOrderedBuilder) {
+		r.dispatchThrottlingCheckResolverEnabled = enabled
+		r.dispatchThrottlingCheckResolverOptions = opts
+	}
+}
+
+func NewOrderedCheckResolvers(opts ...CheckResolverOrderedBuilderOpt) *CheckResolverOrderedBuilder {
+	checkResolverBuilder := &CheckResolverOrderedBuilder{}
+	checkResolverBuilder.resolvers = []CheckResolver{}
+	for _, opt := range opts {
+		opt(checkResolverBuilder)
+	}
+	return checkResolverBuilder
+}
+
+// Build constructs a CheckResolver that is composed of various CheckResolvers in the manner of a circular linked list.
+// The resolvers should be added from least resource intensive to most resource intensive.
+// CycleDetectionCheckResolver is always the first resolver in the composition and LocalChecker is always the last.
+//
+//	CycleDetectionCheckResolver  <----------------------|
+//		[...Other resolvers depending on the opts order]
+//			LocalChecker    ----------------------------^
+//
+// The returned CheckResolverCloser should be used to close all resolvers involved in the list.
+func (c *CheckResolverOrderedBuilder) Build() (CheckResolver, CheckResolverCloser) {
```

Comment:
thought: what do you think about having this function just be part of the constructor logic? You'll always call the constructor followed by Build. Calling Build more than once on an instance likely would yield incorrect behavior right now.

Reply:
Good catch, `Build` is not idempotent.

The issue with merging them is that we won't be able to write unit tests that assert on the value of `builder.resolvers` 🤔